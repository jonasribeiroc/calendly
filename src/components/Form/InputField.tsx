import React from 'react';
import styled from 'styled-components';
import InputLabel from './InputLabel';

interface InputFieldProps {
    id: string;
    label: string;
    type?: string;
    required?: boolean;
    value: string;
    onChange: (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => void;
    placeholder?: string;
    hasError?: boolean;
    errorMessage?: string;
}

const InputContainer = styled.div`
    display: flex;
    flex-direction: column;
    margin-bottom: 20px;
`;

interface StyledInputProps {
    $hasError?: boolean
}

const StyledInput = styled.input<StyledInputProps>`
    padding: 12px;
    border: ${(props) => (props.$hasError ? '2px solid red' : '1px solid #ccc')};
    border-radius: 4px;
    font-size: 15px;

    &:focus {
        outline: none;
        border-color: #007bff;
    }
`;

const ErrorMessage = styled.span`
    color: red;
    font-size: 0.875rem;
`;

const InputField: React.FC<InputFieldProps> = ({
    id,
    label,
    type = 'text',
    required = false,
    value,
    onChange,
    placeholder = '',
    hasError = false,
    errorMessage = '',
}) => {
    return (
        <InputContainer>
            <InputLabel htmlFor={id} label={label} required={required} />
            <StyledInput
                id={id}
                type={type}
                required={required}
                value={value}
                onChange={onChange}
                placeholder={placeholder}
                $hasError={hasError}
            />
            {hasError && <ErrorMessage>{errorMessage}</ErrorMessage>}
        </InputContainer>
    );
};

export default InputField;
