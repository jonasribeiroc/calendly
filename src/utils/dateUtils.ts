import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
    getDay,
    lastDayOfMonth,
    isSameMonth as isSameMonthFns,
    startOfMonth,
    endOfMonth,
    getDate,
} from 'date-fns';

function getDateUTC(date?: Date): Date {
    const baseDate = date instanceof Date ? date : new Date();

    return new Date(Date.UTC(
        baseDate.getUTCFullYear(),
        baseDate.getUTCMonth(),
        baseDate.getUTCDate(),
        baseDate.getUTCHours(),
        baseDate.getUTCMinutes(),
        baseDate.getUTCSeconds(),
        baseDate.getUTCMilliseconds()
    ));
}

function isSameMonth(date1: Date, date2: Date): boolean {
    return isSameMonthFns(date1, date2);
}

function isValidDateString(str: string, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str, formatStr, getDateUTC());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDateFromString(str: string | null, dateFormat: string): Date | undefined {
    if (str && isValidDateString(str, [dateFormat])) {
        return getDateUTC(new Date(str.replace('-', ',')));
    }
}

function getDateString(date: Date, formatStr?: string): string {
    return format(date, formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    return getDate(lastDayOfMonth(date));
}

function getWeekDay(date: Date): number {
    return getDay(date);
}

function getWeekDayName(date: Date): string {
    return format(date, 'EEEE');
}

function getMonthDay(date: Date): number {
    return Number(format(date, 'd'));
}

function getTime(date: Date): string {
    return format(date, 'HH:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    const firstDay = startOfMonth(date);

    return getDateUTC(new Date(getYear(firstDay), getMonth(firstDay), getMonthDay(firstDay), 0, 0, 0, 0));
}

function getLastDayOfMonth(date: Date): Date {
    const lastDay = endOfMonth(date);

    return getDateUTC(new Date(getYear(lastDay), getMonth(lastDay), getMonthDay(lastDay), 23, 59, 59, 999));
}

function getAvailableDays(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameMonth(d, date))
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameMonth,
    isValidDateString,
    getDateUTC,
    getDateFromString,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getMonthDay,
    getTime,
    getFirstDayOfMonth,
    getLastDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}