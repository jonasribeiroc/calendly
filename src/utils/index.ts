import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
} from 'date-fns';

function isSameYearMonth(str1?: string | null, str2?: string | null): boolean {
    if (!str1 || !str2) return false;

    const [year1, month1] = str1.split('-');
    const [year2, month2] = str2.split('-');

    return year1 === year2 && month1 === month2;
}

function isValidDateStr(str: string | null, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str || '', formatStr, new Date());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDate(str?: string | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd'): Date {
    return !!str && isValidDateStr(str, ['yyyy-MM-dd', 'yyyy-MM']) ? parse(str, formatStr || 'yyyy-MM', new Date()) : new Date();
}

function getDateString(date?: Date | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd' | 'yyyy-MM-dd\'T\'HH:mm:ss' | "yyyy-MM-dd'T'HH:mm:ssxxx"): string {
    return format(date || new Date(), formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    const year = getYear(date);
    const month = getMonth(date);

    return new Date(year, month, 0).getDate();
}

function getWeekDay(date: Date): number {
    return date.getUTCDay();
}

function getWeekDayName(date: Date): string {
    return date.toLocaleString('en-US', { weekday: 'long' });
}

function getDay(date: Date): number  {
    return Number(date.toLocaleString('en-US', { day: 'numeric' }));
}

function getTime(date: Date): string {
    return format(date, 'hh:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    return new Date(date.getFullYear(), date.getMonth(), 1);
  }

function getAvailableDays(availableDateTimes: string[], date: Date): number[] {
    const daysSet = new Set<number>();
    const year = getYear(date);
    const month = getMonth(date);

    availableDateTimes.forEach((dateTime) => {
        const date = new Date(dateTime);
        const dateYear = getYear(date);
        const dateMonth = getMonth(date);

        if (dateYear === year && dateMonth === month) {
            daysSet.add(getDay(date));
        }
    });

    return Array.from(daysSet).sort((a, b) => a - b);
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameYearMonth,
    isValidDateStr,
    getDate,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getDay,
    getTime,
    getFirstDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}