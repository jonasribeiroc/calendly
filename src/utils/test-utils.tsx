import React, { ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

interface AllTheProvidersProps {
  children: ReactNode;
}

const AllTheProviders = ({ children }: AllTheProvidersProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
