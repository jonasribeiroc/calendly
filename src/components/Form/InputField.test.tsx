import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import { InputField } from './InputField';
import { render } from '../../utils/test-utils';

describe('InputField Component', () => {
    const mockOnChange = jest.fn();
    const mockId = 'test-input';
    const mockLabel = 'Test Label';
    const mockValue = 'Test Value';
    const mockPlaceholder = 'Enter text';
    const mockErrorMessage = 'This field is required';

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders the label with correct text', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
            />
        );
        expect(screen.getByText(mockLabel)).toBeInTheDocument();
    });

    it('associates the label with the correct htmlFor', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
            />
        );
        const label = screen.getByText(mockLabel);
        expect(label).toHaveAttribute('for', mockId);
    });

    it('renders the input with correct attributes', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                type="email"
                value={mockValue}
                onChange={mockOnChange}
                placeholder={mockPlaceholder}
            />
        );
        const input = screen.getByPlaceholderText(mockPlaceholder) as HTMLInputElement;
        expect(input).toBeInTheDocument();
        expect(input).toHaveAttribute('type', 'email');
        expect(input).toHaveAttribute('id', mockId);
        expect(input).toHaveValue(mockValue);
    });

    it('shows asterisk when required is true', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                required
                value={mockValue}
                onChange={mockOnChange}
            />
        );
        expect(screen.getByText('*')).toBeInTheDocument();
    });

    it('does not show asterisk when required is false', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
            />
        );
        expect(screen.queryByText('*')).not.toBeInTheDocument();
    });

    it('calls onChange when input value changes', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value=""
                onChange={mockOnChange}
                placeholder={mockPlaceholder}
            />
        );
        const input = screen.getByPlaceholderText(mockPlaceholder) as HTMLInputElement;
        fireEvent.change(input, { target: { value: 'New Value' } });
        expect(mockOnChange).toHaveBeenCalledTimes(1);
        expect(mockOnChange).toHaveBeenCalledWith(expect.any(Object));
    });

    it('applies error styles when hasError is true', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
                hasError
            />
        );
        const input = screen.getByDisplayValue(mockValue);
        expect(input).toHaveStyle('border: 2px solid #e74c3c');
    });

    it('renders error message when hasError is true', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
                hasError
                errorMessage={mockErrorMessage}
            />
        );
        expect(screen.getByText(mockErrorMessage)).toBeInTheDocument();
    });

    it('does not render error message when hasError is false', () => {
        render(
            <InputField
                id={mockId}
                label={mockLabel}
                value={mockValue}
                onChange={mockOnChange}
                hasError={false}
                errorMessage={mockErrorMessage}
            />
        );
        expect(screen.queryByText(mockErrorMessage)).not.toBeInTheDocument();
    });
});
