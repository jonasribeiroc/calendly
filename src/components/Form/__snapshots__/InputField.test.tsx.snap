// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField Component matches the snapshot with error 1`] = `
<DocumentFragment>
  <div
    class="sc-eaUbBy fdTNpf"
  >
    <label
      class="sc-KXCwU cDImbL"
      for="test-input"
    >
      Test Label
    </label>
    <input
      class="sc-gaZyOd dLOVyS"
      id="test-input"
      placeholder=""
      type="text"
      value="Test Value"
    />
    <span
      class="sc-dDvxFM dxUBU"
    >
      This field is required
    </span>
  </div>
</DocumentFragment>
`;

exports[`InputField Component matches the snapshot without error 1`] = `
<DocumentFragment>
  <div
    class="sc-eaUbBy fdTNpf"
  >
    <label
      class="sc-KXCwU cDImbL"
      for="test-input"
    >
      Test Label
    </label>
    <input
      class="sc-gaZyOd liDFrS"
      id="test-input"
      placeholder="Enter text"
      type="text"
      value="Test Value"
    />
  </div>
</DocumentFragment>
`;
